client/src/lib.rs

```diff
@@ -124,118 +136,72 @@ impl SubxtClient {
-                                params,
+                                raw,
+                                subscribe_id,
+                                unsubscribe_id,
                                 unsubscribe_method,
                                 send_back,
                             }) => {
-                                {
-                                    let mut subscriptions = subscriptions.write().await;
-                                    subscriptions.insert(request_id, unsubscribe_method);
-                                }
-
-                                let request =
-                                    Request::Single(Call::MethodCall(MethodCall {
-                                        jsonrpc: Version::V2,
-                                        method: subscribe_method,
-                                        params,
-                                        id: Id::Num(request_id),
-                                    }));
+                                let raw_response = rpc.rpc_query(&session, &raw).await;
+                                let sub_id: SubscriptionId = match read_jsonrpc_response(
+                                    raw_response,
+                                    Id::Number(subscribe_id),
+                                ) {
+                                    Some(Ok(rp)) => {
+                                        serde_json::from_value(rp)
+                                            .expect("infalliable; qed")
+                                    }
+                                    Some(Err(e)) => {
+                                        send_back
+                                            .send(Err(e))
+                                            .expect("failed to send request response");
+                                        return
+                                    }
+                                    None => return,
+                                };
 
                                 let (mut send_front_sub, send_back_sub) =
                                     mpsc::channel(DEFAULT_CHANNEL_SIZE);
-                                if let Ok(message) = serde_json::to_string(&request) {
-                                    if let Some(response) =
-                                        rpc.rpc_query(&session, &message).await
-                                    {
-                                        let result = match serde_json::from_str::<Output>(
-                                            &response,
-                                        )
-                                        .expect("failed to decode subscription response")
-                                        {
-                                            Output::Success(_) => {
-                                                Ok((
-                                                    send_back_sub,
-                                                    SubscriptionId::Num(request_id),
```

Comment:
This was wrong and fixed, it would use the actual request ID. Thus dropping the subscription wouldn't unless the server had configured the actual requestID as subscription ID :)